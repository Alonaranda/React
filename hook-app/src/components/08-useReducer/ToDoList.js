import React from "react";
import { TodoListItem } from "./TodoListItem";

export const ToDoList = ({ toDos, handleDelete, handleToggle }) => {
  return (
    <ul className="list-group list-group-flush">
      {toDos.map((toDo, i) => (
          <TodoListItem
            key={toDo.id}
            toDo={toDo}
            i={i}
            handleDelete={handleDelete}
            handleToggle={handleToggle}
          />
      ))}
    </ul>
  );
};

