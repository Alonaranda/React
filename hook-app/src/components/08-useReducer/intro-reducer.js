
const initialState = [{
    id: 1,
    toDo: 'Comprar pan',
    done: false
}];

const toDoReducer = (state = initialState, action) => {
    //Dentro de aqui tengo que hacer algo para modificar el state
    if(action?.type === 'agregar'){
        return [...state, action.payload];
    }

    return state;
}

let toDos = toDoReducer();

//No usar push en React, solo en useRef
const newToDo = {
    id: 2,
    toDo: 'Comprar leche',
    done: false
}
const agregarToDoAction = {
    type: 'agregar',
    payload: newToDo
}

toDos = toDoReducer(toDos, agregarToDoAction);











console.log(toDos);