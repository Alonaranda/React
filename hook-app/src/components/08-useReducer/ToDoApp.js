import React, {useReducer, useEffect} from 'react';
import { toDoReducer } from './toDoReducer';
import {useForm} from '../hooks/useForm'
import './style.css';
import { ToDoList } from './ToDoList';


//Funcion init
const init = () => {
    /*
    return [{
        id: new Date().getTime(),
        descripcion: 'Aprender React',
        done: false
    }];*/
    return JSON.parse(localStorage.getItem('todos')) || [];
}


export const ToDoApp = () => {

    //Usamos useReducer
    const [toDos, dispatch] = useReducer(toDoReducer, [], init);

    //Hooks de forms para extraer los datos
    const [{descripcion}, handleInputChange, reset] = useForm({
        descripcion: '',
    })

    //useEffect para guardar en localStorage cada que toDos cambia
    useEffect(() => {
        localStorage.setItem('todos', JSON.stringify(toDos));
    }, [toDos]);


    //Click en agregar
    const handleSubmit = (e) => {
        e.preventDefault();
        //console.log('Agregar');
        if(descripcion.trim().length <= 1){
            return;
        }
        const newToDo = {
            id: new Date().getTime(),
            descripcion: descripcion,
            done: false
        };
        const action = {
            type: 'add',
            payload: newToDo
        }
        dispatch(action);
        reset();
    }

    //Click en eliminar
    const handleDelete = (id) => {
        //console.log(id);
        const action = {
            type: 'delete',
            payload: id
        }
        dispatch(action);
    }

    const handleToggle = (id) => {
        //console.log('Toogle');
        const action = {
            type: 'toggle',
            payload: id
        }
        dispatch(action);
    }


    return (
        <div>
            <h1>ToDoApp ({toDos.length})</h1>
            <hr/>
            <div className="row">
                <div className="col-7">
                <ToDoList 
                    toDos={toDos} 
                    handleDelete={handleDelete} 
                    handleToggle={handleToggle}
                />
                </div>
                <div className="col-5">
                    <h4>Agregar ToDo</h4>
                    <hr/>

                    <form onSubmit={handleSubmit}>
                        <input
                            type="text"
                            name="descripcion"
                            placeholder="Aprender ..."
                            autoComplete="off"
                            className="form-control"
                            value={descripcion}
                            onChange={handleInputChange}
                        />
                        <button
                            type="submit"
                            className="btn btn-outline-primary btn-block mt-1"
                        >Agregar</button>
                    </form>
                </div>
            </div>
        </div>
    )
}
