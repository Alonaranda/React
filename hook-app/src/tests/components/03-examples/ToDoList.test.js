import {shallow} from 'enzyme';
import React from 'react';
import {demotoDo } from '../../fixtures/demotoDo';
import { ToDoList } from '../../../components/08-useReducer/ToDoList';

describe('Probando toDoList', () => {

    const handleDelete = jest.fn();
    const handleToggle = jest.fn();


    const wrapper = shallow(
    <ToDoList
        toDos={demotoDo}
        handleDelete={handleDelete}
        handleToggle={handleToggle}
    />);


    test('Mostrar correctamente', () => {
        expect(wrapper).toMatchSnapshot();
    });

    test('Debe de tener dos elementos', () => {
        expect(wrapper.find('TodoListItem').length).toBe(demotoDo.length);
        expect(wrapper.find('TodoListItem').at(0).prop('handleDelete')).toEqual(expect.any(Function));
    });
});