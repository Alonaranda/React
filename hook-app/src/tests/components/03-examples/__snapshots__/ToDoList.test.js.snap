// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Probando toDoList Mostrar correctamente 1`] = `
<ul
  className="list-group list-group-flush"
>
  <TodoListItem
    handleDelete={[MockFunction]}
    handleToggle={[MockFunction]}
    i={0}
    key="1"
    toDo={
      Object {
        "descripcion": "Aprender React",
        "done": false,
        "id": 1,
      }
    }
  />
  <TodoListItem
    handleDelete={[MockFunction]}
    handleToggle={[MockFunction]}
    i={1}
    key="2"
    toDo={
      Object {
        "descripcion": "Aprender Mongo",
        "done": false,
        "id": 2,
      }
    }
  />
</ul>
`;
